### ISSUE TITLE
`Oj::Parser#load` fails on buffered input

### ISSUE THREAD

**Issue by Bo98** [ohler55/oj#926]
This code works:

```rb
data = Oj.dump([1] * 8000)
Oj::Parser.usual.load(StringIO.new(data))
```

This code however fails:

```rb
data = Oj.dump([1] * 9000)
Oj::Parser.usual.load(StringIO.new(data))
```

```
parse error, not closed at 1:-1 (EncodingError)
```

It seems the check added in https://github.com/ohler55/oj/commit/0fd31c474702d7daa56cf3be7b568ef33359962d runs on each parse block rather than once when everything is finished.

**Comment by ohler55** [comment]
I'll look into the history of the depth limit. I believe it has something to do with being compatible with the JSON gem but I'll have to look back at old notes.

**Comment by Bo98** [comment]
To be clear, the issue happens when the input is over a certain byte size (16385) regardless of the number of elements or the level of nesting. Another problematic example:

```rb
data = Oj.dump([1000000] * 2049)
Oj::Parser.usual.load(StringIO.new(data))
```

It's when it has to do multiple parses rather than one: https://github.com/ohler55/oj/blob/0fd31c474702d7daa56cf3be7b568ef33359962d/ext/oj/parser.c#L1358-L1365

The issue also does not happen with `Oj.load`, which operates fine with significantly larger inputs:

```rb
data = Oj.dump([1] * 100000000)
Oj.load(StringIO.new(data))
```

Ruby's `JSON.load` also works ok.

**Comment by ohler55** [comment]
That is helpful. Thank you.

**Comment by ohler55** [comment]
Just pushed a branch names "fix-parser-eof" that has the fix. Please let me know if it works for you.

**Comment by Bo98** [comment]
Works great, thanks!

Git diff looks a little weird since it seems like a mixture of tabs and spaces are used for indentation: https://github.com/ohler55/oj/commit/036be24785abfb22d71feb8ffed3ad41f90083d0. But that's just a minor thing.

**Comment by Bo98** [comment]
`load_rescue` already catches `EOFError`s so you could also move the check there instead of a `eof?` check. Both works in the end.

**Comment by ohler55** [comment]
I considered that but to catch the eof error in the C code would require a wrapper function and some way to pass back the result. Now if ruby, most definitely.

**Comment by ohler55** [comment]
Released v3.16.4 with the fix.

**Comment by Bo98** [comment]
Thank you!
